Move changelog from YAML to Markdown

FloEdelmann opened the pull request:
… and automatically convert them in a Gradle task to HTML strings in a JSON file.

* Part of #5415
* See also #3355

westnordost commented:
Oh, that's cool! The markdown also makes it more convenient to write changelogs from now on. The task should be part of the `updateSTreetCompleteData` job, that's the one I execute every time before release.

I wonder if it made more sense to have the changelog simply as a giant HTML and the changelog dialog just cuts off when it finds <h1>$older version number</h1>?

westnordost reviewed (commented):
Did you convert the changelog to markdown by hand or did you write a script? Can you post the source code?

I think it would be slightly more convenient (for adding a new version, but also for reading) if there was just one giant changelog.md source file that gets appended. That changelog.md could then live in the repo root, as is quite common, I think? And if it gets processed anyway for in-app display, it doesn't matter if it is files or one file.
(Though, yesterday night I proposed to keep it as a giant HTML file and not take it apart. I stand to it insofar as I do not think it will affect performance (or maybe even positively), plus, in the changelog fragment, the scrollbar would behave not erratically, which is also a plus)

The script looks fine.

FloEdelmann commented:
> The task should be part of the `updateSTreetCompleteData` job, that's the one I execute every time before release.

Done in [`d112d36`](https://github.com/streetcomplete/StreetComplete/pull/5439/commits/d112d3690e266b272fcb55ed6fb16d7866b91deb).

> Did you convert the changelog to markdown by hand or did you write a script? Can you post the source code?

I wrote a small Node.js script using an [existing module](https://www.npmjs.com/package/node-html-markdown):

```js
#!/usr/bin/env node

import { readFileSync, writeFileSync } from "fs";
import { NodeHtmlMarkdown } from "node-html-markdown";

const changelog = JSON.parse(readFileSync("changelog.json", "utf8"));

for (let [version, changes] of Object.entries(changelog)) {
    changes =
        NodeHtmlMarkdown.translate(changes, {
            bulletMarker: "-",
        }) + "\n";

    writeFileSync(`${version}.md`, changes);
}
```

Then I went over the generated files and adjusted them manually. I found some syntax errors and typos there, which I corrected in the `master` branch in https://github.com/streetcomplete/StreetComplete/commit/425836ad09ecd63eb0c2ad1a26a2edc326842714.

> That changelog.md could then live in the repo root, as is quite common, I think?

You're right, I also like this approach. I'll implement that now.

> Regarding F-Droid changelogs

Thanks for the info. I only linked that issue for reference though; I don't intend to generate those in this PR.

westnordost commented:
> Thanks for the info. I only linked that issue for reference though; I don't intend to generate those in this PR.

Right, okay, I also do not care that much. (But for anyone who does care, it could be easier now)

FloEdelmann commented:
I combined all the Markdown release notes files into one root-level `CHANGELOG.md` file now, which the Gradle task then converts to HTML. Unfortunately, rendering that HTML in a TextView is really bad performance-wise, so I used a WebView instead. Styling that is a bit difficult, but I think I managed to match the previous look.

westnordost commented:
>  Unfortunately, rendering that HTML in a TextView is really bad performance-wise

Oh, that was unexpected. Hm, well, I guess what is most important is that it looks "native", i.e. the font and sizes are the same as in the rest of the app

westnordost commented on buildSrc/src/main/java/UpdateChangelogTask.kt:
```diff
@@ -0,0 +1,42 @@
+    @get:Input lateinit var sourceFile: String
+    @get:Input lateinit var targetFile: String
+
+    private val markdownFlavour = CommonMarkFlavourDescriptor()
```
Can we use Github flavour? I post the current changelog also for each result, so it would be nice if I could use all the markdown that is supported by github

FloEdelmann replied:
Yes, it's easy to switch to `GFMFlavourDescriptor`, but it only changes one thing in the output: `www.AddisMap.com` (from [v33.2](https://github.com/streetcomplete/StreetComplete/releases/tag/v33.2)) is rendered as a link instead of plaintext, because GFM linkifies URLs, while CommonMark does not.

westnordost replied:
That's fine, it's for the future, too

FloEdelmann replied:
Okay, changed to GitHub Flavored Markdown in [`f20d12b`](https://github.com/streetcomplete/StreetComplete/pull/5439/commits/f20d12b5ce78213e9ec9e21844f88aafd726bcf6).

westnordost commented on buildSrc/src/main/java/UpdateChangelogTask.kt:
```diff
@@ -0,0 +1,42 @@
+    @get:Input lateinit var targetFile: String
+
+    private val markdownFlavour = CommonMarkFlavourDescriptor()
+    private val json = Json { prettyPrint = true }
```
unused
```suggestion

```

FloEdelmann replied:
Good catch! Removed in [`51d0b31`](https://github.com/streetcomplete/StreetComplete/pull/5439/commits/51d0b312d1b6e6b02a877b3dc8a81a33d7708f75).

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/screens/about/ChangelogFragment.kt:
```diff
@@ -52,20 +50,21 @@ class WhatsNewDialog(context: Context, sinceVersion: String) : AlertDialog(conte
+            var sinceVersionIndex = fullChangelog.indexOf("<h2>$sinceVersion</h2>")
+            if (sinceVersionIndex == -1) {
+                // if version not found, just show the last one
+                sinceVersionIndex = fullChangelog.indexOf("<h2>", 4)
```
This implementation seems a bit unstable too me. If I understand correctly, it intends to finds the second `<h2>` in the document and does this by leaving the first 4 characters out of the search. This breaks if the HTML changelog file does not immediately start with the header of the most current changelog. (Whitespaces?)

A more stable way would be to find the first occurrence, then search again from the first occurrence. Or, maybe more generally, maybe a helper extension function `fun String.indicesOf(): Sequence<Int>` could help to make it better readable?

FloEdelmann replied:
Yeah, you're right. I like the `indicesOf` idea and implemented it in [`1811ddc`](https://github.com/streetcomplete/StreetComplete/pull/5439/commits/1811ddc410f494c205bb24bef812476cef1100a5).

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/screens/about/ChangelogFragment.kt:
```diff
@@ -75,33 +74,57 @@ class WhatsNewDialog(context: Context, sinceVersion: String) : AlertDialog(conte
+    val verticalMargin = resources.getDimensionInDp(R.dimen.activity_vertical_margin)
+    val horizontalMargin = resources.getDimensionInDp(R.dimen.activity_horizontal_margin)
+
+    val style = """
+        <meta name="color-scheme" content="dark light">
+        <style>
+            body {
+                margin: ${verticalMargin}px ${horizontalMargin}px;
+                color: $textColor;
+                font-size: ${textSize}px;
+            }
+
+            :link { color: $linkColor; }
+
+            h2, h3, h4 { font-family: sans-serif-condensed; }
+
+            h2 { font-size: ${h2Size}px; }
+            h3 { font-size: ${h3Size}px; }
+            h4 { font-size: ${h4Size}px; }
+
+            h2:not(:first-child) {
+                border-top: 1px solid $dividerColor;
+                padding-top: 1rem;
+            }
+
+            @media (prefers-color-scheme: dark) {}
+        </style>
+    """.trimIndent()
```
Too bad that you have to do this extra work to make this look native. If you are already doing that, I think it may be better to actually create a kind of valid html and put the `html` content in `<body></body>` and the above in `<head></head>`?

Anyway, did you test:
- it works fine both on light and dark mode
- text scales with the system setting as other text in the system does, too
?

FloEdelmann replied:
It doesn't seem to make a difference, but creating valid HTML might be a good idea anyway. I changed that in [`41b89e0`](https://github.com/streetcomplete/StreetComplete/pull/5439/commits/41b89e09177669780f74a8c0ffff9dcabd099ec3).

I tested (on Android 13) that the WebView behaves correctly in both light and dark mode and with various text and display size settings, without restarting the app.

westnordost reviewed (commented):
I guess, `TextView` always renders as a whole, so if the text is very large, a lot of it is rendered off-screen and then discarded. `WebView` inevitably needs optimizations for that case, otherwise the web would be just incredibly slow.

Anyway. The resulting CHANGELOG.md looks very clean and will be smooth to extend. You committed some clean-ups of the changelog on master, so I trust that it is fine as is now.

Not too important, but I would like if the resulting changelog html had a comment at the beginning that advises that it should not be edited because it is created automatically from XYZ and if easily possible I would prefer if it was formatted in a human readable manner.

FloEdelmann commented:
> I would like if the resulting changelog html had a comment at the beginning that advises that it should not be edited because it is created automatically from XYZ

Done in [`2c35760`](https://github.com/streetcomplete/StreetComplete/pull/5439/commits/2c35760b560065b092a75e4831ba70441160f674).

> if easily possible I would prefer if it was formatted in a human readable manner

I would also have liked that, but unfortunately it's not easily possible in https://github.com/JetBrains/markdown. We could switch to a different Markdown to HTML converter though.

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/util/ktx/String.kt:
```diff
@@ -4,3 +4,7 @@ fun String.truncate(length: Int): String =
     if (this.length > length) substring(0, length - 1) + "…" else this
 
 fun String.containsAll(words: List<String>) = words.all { this.contains(it) }
+
+/** Returns the indices of all occurrences of [substring]. */
+fun String.indicesOf(substring: String) =
+    generateSequence(indexOf(substring)) { indexOf(substring, it) }
```
Uh, wait a moment... that works?

What about -1? If the substring cannot be found, the sequence should be empty, not consist of unlimited -1s.

And, shouldn't the offset be `it+1`?

westnordost replied:
(using `sequence` with `yield` may result to be more correct and readable, maybe?)

FloEdelmann replied:
You're right, I didn't test it properly. I rewrote the implementation with `yield` and added a unit test: [`99aba20`](https://github.com/streetcomplete/StreetComplete/pull/5439/commits/99aba20c10349a4f1e16bc5cf99aa2a7eeec32bf)

westnordost commented on app/src/main/java/de/westnordost/streetcomplete/screens/about/ChangelogFragment.kt:
```diff
@@ -52,20 +51,21 @@ class WhatsNewDialog(context: Context, sinceVersion: String) : AlertDialog(conte
+            var sinceVersionIndex = fullChangelog.indexOf("<h2>$sinceVersion</h2>")
+            if (sinceVersionIndex == -1) {
+                // if version not found, just show the last one
+                sinceVersionIndex = fullChangelog.indicesOf("<h2>").elementAt(1)
```
throws exception if there is no element at 1?

FloEdelmann replied:
Now it does, see [`99aba20`](https://github.com/streetcomplete/StreetComplete/pull/5439/commits/99aba20c10349a4f1e16bc5cf99aa2a7eeec32bf).

westnordost replied:
Actually, my point was that it should not throw an exception in case there are not two `<h2>` headlines in the file. But well, not so important because there are more than two headlines in the file.

westnordost replied:
And, I guess that was the behavior before, too, so my fault

westnordost commented:
Oh, and, what happens if you click on a link in the webview? It should always be opened in the system browser and never in the WebView itself.

westnordost commented:
Can you gauge when you will be able to finish this? I would like to do a release soon, but want to avoid creating additional work for you by modifying the current changelog.yml.

FloEdelmann commented:
Sorry, I didn't have much time the last days. I'll address your comments now.

> what happens if you click on a link in the webview? It should always be opened in the system browser and never in the WebView itself.

That's exactly what happens (at least on my phone with Android 13).
